### DISCUSSION TITLE
Problem of unknown routing

### DISCUSSION THREAD

**Discussion by Joelgullander** [vercel/next.js#12899] [Help] [answered]
Hey, 

Lets say I use a CMS for resolving slugs before I know the pagetype that should be rendered. For example if a user comes it to /en/newspage, and I fetch my CMS for the slug and it returns a pageType property that I need to render like below:

```
{
  url: "/en/newspage",
  pageType: "newspage"
}
```

There is no way for me to build up a parameterized routingtable at runtime before resolving a route. So my general question is if NextJS has a feature for this or what the best approach would be. One approach that comes to my mind is to only use the index page and fetch the slug from there and conditionally render my pagecomponent myself, is there a better way?

**Comment by leerob** [comment, marked as answer]
I'm not sure that architecture is the best approach. Could you have a specific API for the newspage instead of looping multiple pages into the same API endpoint?

For example, `pages/en/newspage/[page].js` would render all of the news pages and would be able to share that same `newsapge` layout. Then, it could fetch information from your CMS in `getStaticProps` / `getStaticPaths` to build out all of the individual news pages.

**Reply by Joelgullander** [reply]
That would not work in our case because there would be different pagetypes under each subroute. For example `/en/about-us` could be a `ContentPage` while there could be `/en/about-us/cases` that could have its own `PageType` ie `CasePage`. So there is no way of knowing the pagetype based off the slug.

Do you get my point?

**Reply by Joelgullander** [reply]
Maybe a bad example, but what I mean in short is that the editors can choose any pagelayout under any subroute. So different pagetypes can be mixed under the same path.

**Reply by leerob** [reply]
A [catch-all dynamic route](https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes) would allow you to handle a nested structure like that. As for changing out which template is shown, maybe something like this. 

```js
function Page({ layout }) {
  if (layout.pagePage === 'newspage') {
    return <div>Newspage</div>
  }

  return 'Other Page'
}

export async function getStaticProps() {
  const res = await fetch('https://.../your-api')
  const layout = await res.json()

  return {
    props: {
      layout
    }
  }
}
```

**Reply by Joelgullander** [reply]
And then I assume I would be able to lazily load the pagetemplates aswell within that catchall page so that it doesnt affect the chunking. I guess that would work.

**Reply by Joelgullander** [reply]
Thanks @leerob

**Reply by leerob** [reply]
No problem! Does that answer your question?
